### DISCUSSION TITLE
Should some aspects of HW be considered a PD?

### DISCUSSION THREAD

**Discussion by sid-agrawal** [sid-agrawal/OSmosis#9] [Ideas]
When trying to answer the original question, `Which PD has access to the physical page resource?`, we stumbled upon the following line of thinking.

We often say that the kernel (or the highest-privileged software; from now on, I just say kernel) has access to the physical memory resource. However, I would claim that this is no longer true once the MMU is enabled, which happens somewhere in the early boot process. After that, the kernel accesses the physical memory via its virtual addresses.

So, if the kernel also accesses the physical memory via a virtual address, then I would claim that the MMU (or something else in the HW) has access to the physical memory resource. We should treat that thing (i.e., the MMU) as a PD, too.
When the virtual memory is enabled—as per a model transition—the kernel-PD transfers all the physical page resources from the kernel PD to the MMU PD.

What else does this enable us to do? O  why this approach is more elegant:
*  This approach acknowledges that hardware is active and can be considered a PD. 
*  This translates well to devices, too; we need not model devices as passive entities.

One interesting side-effect of this approach is modeling the fact that the kernel can map any physical page to its address space, even though it cannot R/W to physical memory directly. This can be solved by finally exploring the `permissions` attribute on the `hold` edge in the model. 

Let's say that we have the following permission on the physical page resource:
* Read (R)
* Write (W)
* Execute (X)
* Grant (G)

Let's assume the first piece of software that starts on the systems is the kernel, which switches from physical to virtual addresses at some point.

(T* is a timestamp in increasing order.)
```
T0: Kernel has `hold-edge (perm=RWXG)` to all physical pages
T1: Virtual memory enabled
T2: Kernel has `hold-edges (perm=G)` to all physical pages. M U PD has hold edges (perm=RWX) for all physical pages.
```

> Q1. We do not have this notion in the model where a PD can give resources to another PD with a subset of permissions. Do you think we should add it?
> Q2. Do you think this approach makes sense?

## Aside on uKernels
Another thing worth discussing is the ability of a process to manage its own address space in uKernel. 
Even in that case, I would claim that the process only has the following resources:
* RW access to the memory for its page tables. The kernel also mediates that.
* Hold-edge to physical page with `G` perms. 

> This is based on multiple discussions with the team, and this is my attempt to summarize the discussion so that it is not lost in the ether.

Seeking input from @achreto @astevins @p-linh

**Comment by sid-agrawal** [comment]
@shauryapatel1995

**Comment by p-linh** [comment]
Just focusing on some complications with the `hold` edge with permissions, I feel that RSI calculation would have to be updated, since each permission type changes how isolated two PDs are?

Some scenarios where our current RSI calculation would all have the same value, `VP` = Virtual Page:
1. Two PDs only `hold` VP1 with `R` perms
2. Two PDs `hold` VP1 with `RW` perms.
3. Two PDs `hold` VP1 with `W` perms
4. PD1 `hold`s VP1 with `W` perms, while PD2 `hold`s it with `R` perms. 
5. PD1 `hold`s VP1 with `G` perms, while PD2 `hold`s it with `RWX` (the Kernel-PD and MMU-PD example from above)

The PDs in `(1)` are definitely more isolated than in `(2)`? In (3), can these two PDs even mess with each other anymore? Since they can only `W` to the page, and never `R` anything. In `(4)`, PD2 can be corrupted by PD1 (if it reads some garbage written by PD1), but PD1 wouldn't be affected by PD2? In `(5)`, PD1 is also unaffected by whatever PD2 does, however, whoever it grants the page to will be susceptible to issues caused by PD2. 

---
Later edit: After thinking about it some more, modelling devices as PDs does seem nice. Just some stream of thoughts: 
- In the case of MMIO, we could say that the device-PD `hold`s some virtual pages, which could `map` to a physical device register resource or something. For PMIO, the separate I/O address space seems handled by the resource spaces. 
- When devices trigger interrupts, could we relate this at all to the scheduler time-slice (CPU resource)? Although, the PD that handles the interrupt will still be the one that actually runs.
- If we take this approach, will *all*  HW devices be a PD? E.g. the CPU itself is also a PD? Is DRAM a PD?

**Reply by achreto** [reply]
## Devices 

We definitely need to model devices as PDs -- they are active entities in the system.

## Physical Resource PD

> When the virtual memory is enabled—as per a model transition—the kernel-PD transfers all the physical page resources from the kernel PD to the MMU PD.

I don't think this would be accurate. The kernel needs a hold edge to the physical pages to map a virtual page to the physical page. I agree that the MMU, in theory, has access to *almost all* of physical memory -- it needs this in order to do the page table walks. Now, enabling the MMU will effectively:

1. create a new virtual resource space and its resources  (the kernel address space) with hold edges from the kernel
2. create mappings to the physical resources
3. remove permissions of the hold edges from the kernel to the physical resources. 

The last point here suggests that we'd need to have another transition. 

## Permissions

> This can be solved by finally exploring the permissions attribute on the hold edge in the model.

Absolutely, this is the next step. We need to think about what permissions we want. I think there are two categories:
1. Use access: Read access (I'm also including execute here), . Write access, i.e., modify the resource
2. Meta access used to somehow obtain either read and/or write access. 

Point 2) is important as this together with the request edges may encode what resources one could obtain use access to. 

In any case, this is a great *refinement* exercise. Which will add new transitions and introduce permissions that are not present in the most abstract model. ;) 

## RSI

Yes, we definitely need to refine our metrics here. 

## Interrupts

That's a very good point. We definitely need to include this!
Some resource: https://retoachermann.ch/static/papers/humbel-2017-tci.pdf

**Comment by p-linh** [comment]
As I'm thinking about what may need to be specified in a PD-configuration to start a VM, I've thought of something that convinces me it makes a lot of sense to model parts of HW as PDs.  

Our current PD-configuration always creates new PDs based off of a "reference" PD, so any resources given to the child PD must always exist in the reference PD. I was trying to think about how to deal with device regions for the VM, as in vanilla sel4test, we just arbitrarily specify a physical address for MMIO seemingly "out of thin air". If devices are modelled as PDs, the PD-configuration for a VM can map in a frame for a device region using the device-PD as the reference PD, specifying that we want to create a VMR for the device region in the VM-PD's address space based off of the device-PD. I feel like this makes a lot more sense compared to just pulling a physical address range out of nowhere. I think this can also extend to driver-PDs as well.
